csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/common/TranslatedCallBlueprint.qll

```diff
@@ -0,0 +1,217 @@
+        result = getFirstArgumentOrCallInstruction()
+      )
+    )
+  }
+
+  override Instruction getInstructionOperand(InstructionTag tag,
+      OperandTag operandTag) {
+    (
+      tag = CallTag() and
+      (
+        (
+          operandTag instanceof CallTargetOperandTag and
+          result = getInstruction(CallTargetTag())
+        ) or
+        (
+          operandTag instanceof ThisArgumentOperandTag and
+          result = getQualifierResult()
+        ) or
+        exists(PositionalArgumentOperandTag argTag |
+          argTag = operandTag and
+          result = getArgument(argTag.getArgIndex()).getResult()
+        )
+      )
+    ) or
+    (
+      tag = CallSideEffectTag() and
+      hasSideEffect() and
+      operandTag instanceof SideEffectOperandTag and
+      result = getUnmodeledDefinitionInstruction()
+    )
+  }
+
+  override final Type getInstructionOperandType(InstructionTag tag,
+      TypedOperandTag operandTag) {
+    tag = CallSideEffectTag() and
+    hasSideEffect() and
+    operandTag instanceof SideEffectOperandTag and
+    result instanceof Language::UnknownType
+  }
+
+  Instruction getResult() {
+    result = getInstruction(CallTag())
+  }
+
+  /**
+   * Gets the result type of the call.
+   */
+  abstract Type getCallResultType();
+
+  /**
+   * Gets the unmodeled definition instruction of the enclosing
+   * function (of the element this call is attached to).
+   */
+  abstract Instruction getUnmodeledDefinitionInstruction();
+  
+  /**
+   * Holds if the call has a `this` argument.
+   */
+  predicate hasQualifier() {
+    exists(getQualifier())
+  }
+
+  /**
+   * Gets the expr for the qualifier of the call. 
+   */
+  abstract TranslatedExprBlueprint getQualifier();
+
+  /**
+   * Gets the instruction whose result value is the `this` argument of the call.
+   * In general, this is just the result of `getQualifier()`, but it can be
+   * overridden by a subclass for cases where there is a `this` argument that is
+   * not computed from a child expression (e.g. a constructor call).
+   */
+  abstract Instruction getQualifierResult();
+
+  /**
+   * Gets the argument with the specified `index`. Does not include the `this`
+   * argument. We use `TranslatedExprBlueprint` so that we can give both `TranslatedExpr` args,
+   * in the case of AST generated arguments, or `TranslatedCompilerElement` args in the case of
+   * compiler generated arguments.
+   */
+  abstract TranslatedExprBlueprint getArgument(int index);
+
+  /**
+   * If there are any arguments, gets the first instruction of the first
+   * argument. Otherwise, returns the call instruction.
+   */
+  final Instruction getFirstArgumentOrCallInstruction() {
+    if hasArguments() then
+      result = getArgument(0).getFirstInstruction()
+    else
+      result = getInstruction(CallTag())
+  }
+
+  /**
+   * Holds if the call has any arguments, not counting the `this` argument.
+   */
+  abstract predicate hasArguments();
```

Comment:
Can't this be implemented in terms of `getArgument()`?

Reply:
Indeed, modified the code.